Ignore allow_suspicious_primary_key on ATTACH and verify on ALTER

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Ignore `allow_suspicious_primary_key` on `ATTACH` and verify on `ALTER`

Follow-up for: https://github.com/ClickHouse/ClickHouse/pull/61399 (cc @tavplubix )

azat committed:
Ignore allow_suspicious_primary_key on ATTACH and verify on ALTER

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
Co-authored-by: Alexander Tokmakov <tavplubix@clickhouse.com>

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #64202 from azat/ClickHouse:allow_suspicious_primary_key-fixes -->
*This is an automated comment for commit bee827276e66c89e9d9f863b9c9cd1abe491fb8d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64202/bee827276e66c89e9d9f863b9c9cd1abe491fb8d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

tavplubix commented on src/Storages/MergeTree/registerStorageMergeTree.cpp:
```diff
@@ -678,8 +667,8 @@ static StoragePtr create(const StorageFactory::Arguments & args)
             args.storage_def->order_by->ptr(), metadata.columns, context, merging_param_key_arg);
-        if (!local_settings.allow_suspicious_primary_key)
-            verifySortingKey(metadata.sorting_key);
+        if (!local_settings.allow_suspicious_primary_key && !attach)
```
```suggestion
        if (!local_settings.allow_suspicious_primary_key && args.mode <= LoadingStrictnessLevel::CREATE)
```

tavplubix commented on src/Storages/MergeTree/registerStorageMergeTree.cpp:
```diff
@@ -792,8 +781,8 @@ static StoragePtr create(const StorageFactory::Arguments & args)
             = KeyDescription::getSortingKeyFromAST(engine_args[arg_num], metadata.columns, context, merging_param_key_arg);
-        if (!local_settings.allow_suspicious_primary_key)
-            verifySortingKey(metadata.sorting_key);
+        if (!local_settings.allow_suspicious_primary_key && !attach)
```
```suggestion
        if (!local_settings.allow_suspicious_primary_key && args.mode <= LoadingStrictnessLevel::CREATE)
```

azat commented:
@tavplubix I've rebased on top of the https://github.com/ClickHouse/ClickHouse/pull/64204 (to fix build failures) and I squashed your commits (but added co-authored tag).

clickgapai commented:
Hi @azat @tavplubix — while reviewing this PR I found the following:

- **Bug (P2):** Settings-only ALTER on `MergeTree` with `SimpleAggregateFunction` in ORDER BY fails unexpectedly — https://github.com/ClickHouse/ClickHouse/issues/104463

Happy to discuss — close anything that's wrong or already addressed.
